### PR TITLE
Diagnostics CLI: return different exit code when only warnings

### PR DESCRIPTION
**Pull request by tennox**
Thanks for this amazing project! :heart: 

Would love to be able to ignore warnings for the [pre-commit hook in my devenv](https://github.com/cachix/pre-commit-hooks.nix/blob/master/modules/hooks.nix#L774)

### CODE REVIEW

**Review by oxalica — CHANGES_REQUESTED** [review]

File: crates/nil/src/main.rs [hunk 1]
```diff
@@ -129,11 +129,19 @@ fn main_diagnostics(args: DiagnosticsArgs) {
 
         let mut writer = StandardStream::stdout(ColorChoice::Auto);
         emit_diagnostics(path, &src, &mut writer, &mut diags.iter().cloned())?;
-        Ok(diags.is_empty())
+
+        Ok(diags.iter().map(|diag| diag.severity())/*.unique()*/.collect())
```
[the remark this answers is not in the document]
Comment (tennox) [reply crates/nil/src/main.rs]: Done :+1:

File: crates/nil/src/main.rs [hunk 2]
```diff
@@ -129,11 +129,19 @@ fn main_diagnostics(args: DiagnosticsArgs) {
...
     })();
     match ret {
-        Ok(true) => {}
-        Ok(false) => process::exit(1),
+        Ok(severities) => {
+            if severities.is_empty() {
```
[the remark this answers is not in the document]
Comment (tennox) [reply crates/nil/src/main.rs]: Agreed, checked [eslint for comparison](https://eslint.org/docs/latest/use/command-line-interface#exit-codes):

>     0: Linting was successful and there are no linting errors. If the [--max-warnings](https://eslint.org/docs/latest/use/command-line-interface#--max-warnings) flag is set to n, the number of linting warnings is at most n.
>     1: Linting was successful and there is at least one linting error, or there are more linting warnings than allowed by the --max-warnings option.
>     2: Linting was unsuccessful due to a configuration problem or an internal error.

I felt `--max-warnings` weird, so i decided for a simple `--error-on-warnings` and implemented it as in your suggestion as a breaking change :tipping_hand_man:  :+1:

File: docs/features.md [hunk 3]
```diff
@@ -120,5 +120,5 @@ You can run `nil --help` for usages of all available commands.
 
 - `nil diagnostics <PATH>`
   Check and print diagnostics for a file.
-  Exit with non-zero code if there are any diagnostics.
+  Exit with non-zero code if there are any diagnostics. (`1` for errors, `2` if only warnings)
```
Comment (oxalica) [inline docs/features.md:123]: This is outdated.

File: crates/nil/src/main.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (oxalica) [reply crates/nil/src/main.rs]: > I felt `--max-warnings` weird,

I don't like it either. I prefer rust's `--{allow,warn,deny} {LINT,warnings}` approach which also supports promoting, ignoring and filtering. With that customization support, we can simply stick on just exit on errors. But we don't need to implement these flags in this PR, whose main goal is to loosen the exit code for warnings. You can leave these flags to me later.

> as in your suggestion as a breaking change

I don't consider accepting more code a breaking change.

**Review by oxalica — CHANGES_REQUESTED** [review]
Also please squash and edit your commit messages to comply to the main branch.

**Comment by tennox** [comment]
Okay, I did the requested changes & hope the squashed commit message is good enough :sweat_smile:

**Comment by oxalica** [comment]
Thanks!

**Pull request merged**